### PR TITLE
Add support for Integer properties and conversion to FIRRTL.

### PR DESCRIPTION
**Pull request by mikeurbach**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes

The Integer property type is defined in FIRRTL spec version 3.1.0: https://github.com/chipsalliance/firrtl-spec/releases/tag/v3.1.0.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x`, `3.6.x`, or `5.x` depending on impact, API modification or big change: `6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by mikeurbach** [comment]
This is an alternative to 3469.

File: core/src/main/scala/chisel3/properties/Property.scala [hunk 1]
```diff
@@ -30,11 +48,7 @@ private[chisel3] object PropertyType {}
   * describe a set of non-hardware types, so they have no width, cannot be used
   * in aggregate Data types, and cannot be connected to Data types.
   */
-class Property[T: PropertyType: TypeTag] extends BaseType {
-
-  /** A TypeTag for the type T of this Property.
-    */
-  private[chisel3] val tpe = typeOf[T]
+class Property[T](implicit propertyType: PropertyType[T]) extends BaseType {
```
Comment (seldridge) [inline core/src/main/scala/chisel3/properties/Property.scala:51]: The sugar way of doing this is:

```suggestion
class Property[T : PropertyType] extends BaseType {
```

File: core/src/main/scala/chisel3/properties/Property.scala [hunk 2]
```diff
@@ -58,6 +72,14 @@ class Property[T: PropertyType: TypeTag] extends BaseType {
...
     clone
   }
+
+  /** Get the IR PropertyType for this Property.
+    *
+    * This delegates to the PropertyType to convert itself to an IR PropertyType.
+    */
+  private[chisel3] def getPropertyType(): ir.PropertyType = {
+    propertyType.getPropertyType()
+  }
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply core/src/main/scala/chisel3/properties/Property.scala:82]: Thanks for the pointer, I had the original sugar on the class definition, but I wasn't sure how to conjure the typeclass instance.

File: core/src/main/scala/chisel3/properties/Property.scala [hunk 3]
```diff
@@ -14,14 +15,31 @@ import scala.annotation.implicitNotFound
   * Chisel.
   */
 @implicitNotFound("unsupported Property type ${T}")
-private[chisel3] class PropertyType[T]
+private[chisel3] trait PropertyType[T] {
+
+  /** Get the IR PropertyType for this PropertyType.
+    */
+  def getPropertyType(): ir.PropertyType
```
Comment (seldridge) [inline core/src/main/scala/chisel3/properties/Property.scala:22]: Kind of a nit: Scala coding conventions treat a 0-arity function declared with parentheses as a function which _has side effects_ and one that does not have parentheses as one which _does not have side effects_. This is admittedly weird. Given that, this should be declared without parentheses.

File: core/src/main/scala/chisel3/properties/Property.scala [hunk 4]
```diff
@@ -14,14 +15,31 @@ import scala.annotation.implicitNotFound
...
 
 /** Companion object for PropertyType.
   *
   * Typeclass instances for valid Property types are defined here, so they will
   * be in the implicit scope and available for users.
   */
-private[chisel3] object PropertyType {}
+private[chisel3] object PropertyType {
+  implicit val intPropertyTypeInstance = new PropertyType[Int] {
+    def getPropertyType(): ir.PropertyType = ir.IntegerPropertyType
```
Comment (seldridge) [inline core/src/main/scala/chisel3/properties/Property.scala:32]: Nit: it's best practice to use the `override` keyword whenever you override an existing method. Scala 2 doesn't require this for whatever reason.

**Review by seldridge — APPROVED** [review]
LGTM

File: core/src/main/scala/chisel3/properties/Property.scala (hunk 3 above)
Comment (mikeurbach) [reply core/src/main/scala/chisel3/properties/Property.scala:22]: ah i was not aware of this convention, thanks!

File: core/src/main/scala/chisel3/properties/Property.scala (hunk 4 above)
Comment (mikeurbach) [reply core/src/main/scala/chisel3/properties/Property.scala:32]: ah yes, makes sense

**Pull request merged**